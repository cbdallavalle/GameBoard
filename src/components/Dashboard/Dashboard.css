.Dashboard {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 30px;
}

nav button {
  font-family: 'Open Sans', sans-serif;
  font-size: 16px;
  background-color: #ffffffcf;
  border-top-left-radius: 13px;
  border-top-right-radius: 13px;
  border: none;
  padding: 10px;
}

nav button:hover,
nav button:focus {
  cursor: pointer;
  background-color: #BD9B89;
}

.Dashboard h2 {
  font-family: 'Righteous', cursive;
  font-size: 25px;
}

nav {
  display: flex;
  flex-direction: row;
  align-items: center;
  width: 630px;
  height: fit-content;
}

.search {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
}

.search button {
  width: 170px;
  height: 42px;
  margin: 0;
  border: 2px solid black;
  font-size: 17px;
  font-family: 'Open Sans', sans-serif;
}

.search input {
  width: 170px;
  height: 35px;
}