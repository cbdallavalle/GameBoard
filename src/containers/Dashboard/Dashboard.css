.Dashboard {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.Dashboard h2 {
  font-family: 'Righteous', cursive;
  font-size: 25px;
}

.body-cont {
  background-color: #f4f4f4a6;
  overflow-y: scroll;
  height: 90vh;
}

.search {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
}

.search button {
  width: 170px;
  height: 42px;
  margin: 0;
  border: 2px solid black;
  font-size: 17px;
  font-family: 'Open Sans', sans-serif;
}

.search input {
  width: 170px;
  height: 35px;
}