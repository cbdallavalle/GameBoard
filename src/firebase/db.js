import { db } from './firebase';

export const doCreateUser = (id, firstName, lastName, email) =>
  db.ref(`users/${id}`).set({
    user: {
      id,
      firstName,
      lastName,
      email,
    }
  });

export const onceGetUsers = async () => {
  try {
    const snapshot = await db.ref('users').once('value');
    if (snapshot.status < 300) {
      const value = await snapshot.val();
      return value
    } else {
      throw new Error('unable to fetch users')
    }
  } catch (error) {
    throw new Error('unable to fetch users')
  }
}

export const doWriteFavoriteData = async (userId, newFavorites) => {
  try {
    db.ref('users/' + userId + '/favorites').set(newFavorites);
  } catch (error) {
    throw new Error('unable to write favorite data')
  }
}

export const doAddFavoriteData = async (userId, favorite) => {
  try {
    const existingFavorites = await getFavorites(userId);
    const favoriteToAdd = checkFavoriteDuplication(existingFavorites, favorite);
    doWriteLastFavoritedData(userId, favorite)
    const newFavorites = {...existingFavorites, ...favoriteToAdd }
    doWriteFavoriteData(userId, newFavorites)
  } catch (error) {
    throw new Error('unable to add favorite data')
  }
}

export const doDeleteFavoriteData = async (userId, favorite) => {
  try {
    const existingFavorites = await getFavorites(userId);
    delete existingFavorites[favorite.name]
    doWriteFavoriteData(userId, existingFavorites)
  } catch (error) {
    throw new Error('unable to delete')
  }
}

export const doWriteLastFavoritedData = async (userId, favorite) => {
  db.ref('users/' + userId + '/lastFavorited').set(favorite);
}

export const checkFavoriteDuplication = (existingFavorites, favorite) => {
  const { name, id, description, image, thumbnail } = favorite;
  const review = "No review - click to add one";
  const favoriteToAdd = { [name]: { id, description, image, thumbnail, name, review } }
  return existingFavorites && existingFavorites[favorite.name] ? {} : favoriteToAdd
}

export const doWriteFriendsData = async (userId, friendId) => {
  try {
    const friends = await getFriends(userId);
    if (friends && !friends.includes(friendId)) {
      friends.push(friendId);
      db.ref('users/' + userId + '/friends').set(friends); 
    } else if (!friends) {
      db.ref('users/' + userId + '/friends').set([friendId]); 
    }
  } catch (error) {
    throw new Error ('unable to add friend')
  }
}

export const getFriends = async (userId) => {
  try {
    const value = await onceGetUsers();
    return value[userId].friends
  } catch (error) {
    throw new Error('unable to load friends data')
  }
}

export const getFavorites = async (userId) => {
  try {
    const value = await onceGetUsers();
    return value[userId].favorites;
  } catch (error) {
    throw new Error ('unable to load favorites data')
  }
}

export const getFriendsFavorites = async (userId) => {
  const friends = await getFriends(userId);
  const value = await onceGetUsers();
  if(friends) {
    return friends.reduce( (accu, friendId) => {
      if(value[friendId].lastFavorited) {
        accu[value[friendId].user.firstName] = value[friendId].lastFavorited
      }
      return accu
    }, {} )  
  } else {
    return {}
  }
}

export const doWriteReviewData = async (userId, favorite, review) => {
  try {
    const { name } = favorite;
    const reviewedGame = {...favorite, review}
    const response = await db.ref('users/' + userId + '/favorites/' + name).set(reviewedGame);
    if(response.status < 300) {
      await doWriteLastFavoritedData(userId, reviewedGame)
    } else {
      throw new Error('unable to process review')
    }
  } catch (error) {
    throw new Error('unable to process review')
  }
}