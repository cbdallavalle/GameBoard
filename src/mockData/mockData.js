export const mockData = {
  mockUsers: {
    YRrtMzWo0jNquYjLCQGT8ftqz703:{
      favorites: {}, 
      friends: [], 
      lastFavorited: {}, 
      user: {
        email: "kubla@email.com",
        firstName: "kubla",
        id: "GZZGrEW9aYXzbxrdphXuq04ng6E3",
        lastName: "khan"
      }
    },
    GZZGrEW9aYXzbxrdphXuq04ng6E3:{
      favorites: {}, 
      friends: [], 
      lastFavorited: {}, 
      user: {
        email: "wren@wren.com",
        firstName: "wren",
        id: "GZZGrEW9aYXzbxrdphXuq04ng6E3",
        lastName: "little" 
      }
    }
  },
  mockAllUsersWithoutSearcher: [
    {
      email: "kubla@email.com", 
      firstName: "kubla", 
      id: "GZZGrEW9aYXzbxrdphXuq04ng6E3", 
      lastName: "khan", 
      key: "YRrtMzWo0jNquYjLCQGT8ftqz703"
    }
  ],
  mockAllUsers: [
    {
      email: "kubla@email.com", 
      firstName: "kubla", 
      id: "GZZGrEW9aYXzbxrdphXuq04ng6E3", 
      lastName: "khan", 
      key: "YRrtMzWo0jNquYjLCQGT8ftqz703"
    },
    {
      email: "wren@wren.com",
      firstName: "wren",
      id: "GZZGrEW9aYXzbxrdphXuq04ng6E3",
      lastName: "little" 
    }
  ],
  mockFriends: ['123', '456', '778'],
  mockFavorites: { 'Race for the Galaxy': {game: 'info'} },
  mockFriendsFavorites: {
    kubla: { 'Race for the Galaxy': {game: 'info'} },
    wren: { 'Captain Sonar': {game: 'info'} }
  },
  initialState: {
    firstName: '',
    lastName: '',
    email: '',
    passwordOne: '',
    passwordTwo: '',
    error: null,
    displayCreate: 'login'
  },
  mockLogInEvent: {
    target: {
      name: 'email', 
      value: 'coda@gmail.com'
    }, 
    preventDefault: () => {}
  },
  mockDefaultSearchState: {
    games: [],
    search: '',
    game: {},
    error: '',
    loading: false,
    gameAdded: 'not-added'
  },
  mockSearchEvent: {
    target: {
      name: 'search',
      value: 'Meeple Circus'
    },
    preventDefault: () => {}
  }, 
  mockSearch: [{ 
    id: '123', 
    name: 'Mysterium' 
  }],
  mockSearchGame: { 
    id: '123', 
    name: 'Mysterium', 
    thumbnail: 'src', 
    image: 'src', 
    description: 'A mystery game with pictures' 
  },
  mockMSTPUserState: {
    user: {
      uid: 'GZZGrEW9aYXzbxrdphXuq04ng6E3'
    }
  },
  mockMSTPUser: {
    user: {
      uid: 'GZZGrEW9aYXzbxrdphXuq04ng6E3'
    }
  }
};