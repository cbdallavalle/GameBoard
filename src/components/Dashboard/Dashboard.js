import React, { Component } from 'react';
import { CardContainer } from '../CardContainer/CardContainer';
import { searchGames } from '../../actions';
import { connect } from 'react-redux';
import { Nav } from '../Nav/Nav';
import Search from '../Search/Search';
import AddFriends from '../AddFriends/AddFriends';
import { SignOutBtn } from '../SignOutBtn/SignOutBtn';
import './Dashboard.css';

export class Dashboard extends Component {
  constructor() {
    super()
    this.state = {
      display: 'your-games',
    }
  }

  determineDisplay = () => {
    if(this.state.display === 'your-games') {
      return (
        <CardContainer favorites={this.props.favorites}/>
      )  
    } else if (this.state.display === 'friends-games') {
      return (
        <CardContainer favorites={{}}/>
      )
    } else if (this.state.display === 'search-games') {
      return (
        <Search />
      )
    } else if (this.state.display === 'search-friends') {
      return (
        <AddFriends />
      )
    }

  }

  handleClick = e => {
    this.setState({ display: e.target.name})
  }

  render() {
    return (
      <section className="Dashboard">
        <nav>
          <button name='your-games' onClick={this.handleClick}>
            <i className="fas fa-user"></i> Your Games
          </button>
          <button name='friends-games' onClick={this.handleClick}>
            <i className="fas fa-users"></i> Your friend's games
          </button>
          <button name='search-games' onClick={this.handleClick}>
            Search for games 
          </button>
          <button name="search-friends" onClick={this.handleClick}>
            Search for friends
          </button>
        </nav>
        {this.determineDisplay()}
      </section>
    )
  }
}

export const mapStateToProps = state => ({
  favorites: state.favorites
})

export default connect(mapStateToProps, null)(Dashboard);