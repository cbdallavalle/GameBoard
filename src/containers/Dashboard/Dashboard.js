import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Switch, Route, withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { db } from '../../firebase';

import AddFriends from '../AddFriends/AddFriends';
import Search from '../Search/Search';
import { CardContainer } from '../../components/CardContainer/CardContainer';
import { Header } from '../../components/Header/Header';
import { Nav } from '../../components/Nav/Nav';

import './Dashboard.css';

export class Dashboard extends Component {
  constructor() {
    super();
    this.state = {
      friendsFavorites: {},
      error: ''
    };
  }

  componentWillReceiveProps = () => {
    this.updateFavorites();
  }

  updateFavorites = async() => {
    try {
      if (this.props.user.uid) {
        const friendsFavorites = await 
          db.getFriendsFavorites(this.props.user.uid);
        this.setState({ friendsFavorites });
      }
    } catch (error) {
      this.setState({ error: error.message });
    }
  }

  render() {
    return (
      <section className="Dashboard">
        <Header />
        <div className="body-cont">
        <Nav handleClick={this.handleClick}/>
          <Switch>
            <Route
              exact path="/dashboard/your-games"
              render={ () => 
                <CardContainer 
                  favorites={this.props.favorites} 
                  type={"games"} 
                /> 
              }
            />
            <Route 
              exact path="/dashboard/friends-games" 
              render={ () => 
                <CardContainer 
                  favorites={this.state.friendsFavorites} 
                  type={"friends"} 
                /> 
              }
            />
            <Route 
              exact path="/dashboard/search-games" 
              component={ Search }
            />
            <Route 
              exact path="/dashboard/search-users" 
              component={ AddFriends }
            />
          </Switch>
        </div>
      </section>
    );
  }
}

Dashboard.propTypes = {
  favorites: PropTypes.object.isRequired,
  user: PropTypes.object
};

export const mapStateToProps = state => ({
  favorites: state.favorites,
  user: state.user
});

export default withRouter(connect(mapStateToProps, null)(Dashboard));