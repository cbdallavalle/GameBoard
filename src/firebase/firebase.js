import * as firebase from 'firebase';

const config = {
  apiKey: "AIzaSyBDU0_Pa8xdAWKmtKKfo3QkbrtR56cl024",
  authDomain: "dallavalle-gameboard.firebaseapp.com",
  databaseURL: "https://dallavalle-gameboard.firebaseio.com",
  projectId: "dallavalle-gameboard",
  storageBucket: "",
  messagingSenderId: "250758219123"
};

if(!firebase.apps.length) {
  firebase.initializeApp(config)
}

const auth = firebase.auth();

export {
  auth,
}