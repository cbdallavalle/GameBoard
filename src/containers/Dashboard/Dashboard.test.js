import React from 'react';
import { shallow } from "enzyme";
import { Dashboard, mapStateToProps } from './Dashboard';
import { db } from '../../firebase';
import { mockData } from '../../mockData/mockData';

describe('Dashboard', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(f
      <Dashboard 
        favorites={{}}
        user={{ uid: '123' }}
      />
    )
  })

  it('should match the snapshot', () => {

    expect(wrapper).toMatchSnapshot();
  })

  it('should have a default state of an empty object for friendsFavorites and error as an empty string', () => {

    expect(wrapper.state()).toEqual({ friendsFavorites: {}, error: '' })
  })

  //updateFavorites
  it('updateFavorites should set state with an error if an error is caught', async() => {

    await wrapper.instance().updateFavorites();
    expect(wrapper.state().error).toEqual('unable to load friends data')
  })

  it('updateFavorites will set friendsFavorites to an array of friends', async() => {

    db.getFriendsFavorites = () => mockData.mockFriendsFavorites;
    await wrapper.instance().updateFavorites();
    expect(wrapper.state().friendsFavorites).toEqual(mockData.mockFriendsFavorites )
  })
})

describe("mapStateToProps", () => {
  const initialState = {
    favorites: mockData.mockFavorites,
    user: mockData.mockMSTPUser.user
  }

  const expected = {
    favorites: mockData.mockFavorites,
    user: mockData.mockMSTPUser.user
  }

  it('mapStateToProps should take in state and return an object', () => {
    expect(mapStateToProps(initialState)).toEqual(expected);
  })
})