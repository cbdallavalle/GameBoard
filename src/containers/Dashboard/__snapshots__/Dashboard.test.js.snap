// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dashboard
    favorites={Object {}}
    user={
        Object {
            "uid": "123",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Header />,
        <div
          className="body-cont"
>
          <Unknown
                    handleClick={undefined}
          />
          <Switch>
                    <Route
                              exact={true}
                              path="/dashboard/your-games"
                              render={[Function]}
                    />
                    <Route
                              exact={true}
                              path="/dashboard/friends-games"
                              render={[Function]}
                    />
                    <Route
                              component={[Function]}
                              exact={true}
                              path="/dashboard/search-games"
                    />
                    <Route
                              component={[Function]}
                              exact={true}
                              path="/dashboard/search-users"
                    />
          </Switch>
</div>,
      ],
      "className": "Dashboard",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Unknown
              handleClick={undefined}
/>,
            <Switch>
              <Route
                            exact={true}
                            path="/dashboard/your-games"
                            render={[Function]}
              />
              <Route
                            exact={true}
                            path="/dashboard/friends-games"
                            render={[Function]}
              />
              <Route
                            component={[Function]}
                            exact={true}
                            path="/dashboard/search-games"
              />
              <Route
                            component={[Function]}
                            exact={true}
                            path="/dashboard/search-users"
              />
</Switch>,
          ],
          "className": "body-cont",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "handleClick": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Route
                  exact={true}
                  path="/dashboard/your-games"
                  render={[Function]}
/>,
                <Route
                  exact={true}
                  path="/dashboard/friends-games"
                  render={[Function]}
/>,
                <Route
                  component={[Function]}
                  exact={true}
                  path="/dashboard/search-games"
/>,
                <Route
                  component={[Function]}
                  exact={true}
                  path="/dashboard/search-users"
/>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "exact": true,
                  "path": "/dashboard/your-games",
                  "render": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "exact": true,
                  "path": "/dashboard/friends-games",
                  "render": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "component": [Function],
                  "exact": true,
                  "path": "/dashboard/search-games",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "component": [Function],
                  "exact": true,
                  "path": "/dashboard/search-users",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Header />,
          <div
            className="body-cont"
>
            <Unknown
                        handleClick={undefined}
            />
            <Switch>
                        <Route
                                    exact={true}
                                    path="/dashboard/your-games"
                                    render={[Function]}
                        />
                        <Route
                                    exact={true}
                                    path="/dashboard/friends-games"
                                    render={[Function]}
                        />
                        <Route
                                    component={[Function]}
                                    exact={true}
                                    path="/dashboard/search-games"
                        />
                        <Route
                                    component={[Function]}
                                    exact={true}
                                    path="/dashboard/search-users"
                        />
            </Switch>
</div>,
        ],
        "className": "Dashboard",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Unknown
                handleClick={undefined}
/>,
              <Switch>
                <Route
                                exact={true}
                                path="/dashboard/your-games"
                                render={[Function]}
                />
                <Route
                                exact={true}
                                path="/dashboard/friends-games"
                                render={[Function]}
                />
                <Route
                                component={[Function]}
                                exact={true}
                                path="/dashboard/search-games"
                />
                <Route
                                component={[Function]}
                                exact={true}
                                path="/dashboard/search-users"
                />
</Switch>,
            ],
            "className": "body-cont",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "handleClick": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <Route
                    exact={true}
                    path="/dashboard/your-games"
                    render={[Function]}
/>,
                  <Route
                    exact={true}
                    path="/dashboard/friends-games"
                    render={[Function]}
/>,
                  <Route
                    component={[Function]}
                    exact={true}
                    path="/dashboard/search-games"
/>,
                  <Route
                    component={[Function]}
                    exact={true}
                    path="/dashboard/search-users"
/>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "exact": true,
                    "path": "/dashboard/your-games",
                    "render": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "exact": true,
                    "path": "/dashboard/friends-games",
                    "render": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "component": [Function],
                    "exact": true,
                    "path": "/dashboard/search-games",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "component": [Function],
                    "exact": true,
                    "path": "/dashboard/search-users",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
